### PR TITLE
Set harness error if async_test function returns a value

### LINKED ISSUE #21435: `async_test` with async function produces undesirable results
This was reported on https://groups.google.com/a/chromium.org/forum/#!topic/ecosystem-infra/2dpo_NDYOgc. If you have a test file which passes an `async` function to `async_test`, and that `async` function throws an exception, the sub-test will TIMEOUT and the harness will get an exception.

For example:

```
async_test(async function(t) {
  throw new Error('foo');
}, 'my test');
```

Harness status ERROR and test status TIMEOUT

```
async_test(async function(t) {
  assert_precondition(false, 'foo');
}, 'my other test');
```

Harness status PRECONDITION_FAILED (misleading...?!) and test status TIMEOUT.

In some ways this is working as intended; nothing has told the test object that it is finished, so it is reasonable for it to still be waiting. It is a bit weird that PRECONDITION_FAILED ends up being propagated to the test harness. In either case, however, I cannot see a use for passing an async function to `async_test`, so perhaps we should guard against it?

Naively I think `async_test` could check the return type of `test_obj.step(` and fail in some way if its a Promise, but I don't know testharness.js that well so that may be nonsense.

### PR DESCRIPTION
**Pull request by stephenmcgruer**
Cherry-picks and modifies work done by frehner@ in
https://github.com/web-platform-tests/wpt/pull/24021

Fixes https://github.com/web-platform-tests/wpt/issues/21435

Co-authored-by: Anthony Frehner <afrehner.work@gmail.com>

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
This has revealed another set of tests that return values to `async_test` (all promises):

* [ ] [encrypted-media/clearkey-mp4-reset-src-after-setmediakeys.https.html](https://wpt.fyi/results/encrypted-media/clearkey-mp4-reset-src-after-setmediakeys.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [encrypted-media/drm-mp4-onencrypted.https.html](https://wpt.fyi/results/encrypted-media/drm-mp4-onencrypted.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [encrypted-media/drm-mp4-reset-src-after-setmediakeys.https.html](https://wpt.fyi/results/encrypted-media/drm-mp4-reset-src-after-setmediakeys.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [webrtc/RTCPeerConnection-getStats.https.html](https://wpt.fyi/results/webrtc/RTCPeerConnection-getStats.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [webrtc/RTCPeerConnection-setRemoteDescription-replaceTrack.https.html](https://wpt.fyi/results/webrtc/RTCPeerConnection-setRemoteDescription-replaceTrack.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [webrtc/RTCPeerConnection-track-stats.https.html](https://wpt.fyi/results/webrtc/RTCPeerConnection-track-stats.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)
* [ ] [webrtc/legacy/RTCPeerConnection-addStream.https.html](https://wpt.fyi/results/webrtc/legacy/RTCPeerConnection-addStream.https.html?diff&filter=ADC&q=is%3Adifferent&run_id=702130003&run_id=698170007)

**Comment by stephenmcgruer** [comment]
Latest run on Chrome Dev is good - [no new errors](https://wpt.fyi/results/?diff&filter=ADC&q=is%3Adifferent&run_id=679090004&run_id=688560005).

Since the change should show up for any browser, modulo crashes (as its just looking for the return value of the function given to `async_test`), I'm happy to send this for review now.

File: resources/testharness.js [hunk 1]
```diff
@@ -568,7 +568,21 @@ policies and contribution forms [3].
         var test_name = name ? name : test_environment.next_default_test_name();
         var test_obj = new Test(test_name, properties);
         if (func) {
-            test_obj.step(func, test_obj, test_obj);
+            let value = test_obj.step(func, test_obj, test_obj);
```
Comment (foolip) [inline resources/testharness.js:571]:
```suggestion
            var value = test_obj.step(func, test_obj, test_obj);
```

For consistency with surrounding code.

**Review by foolip — APPROVED** [review]
Sweeeeeeeeeeet!

File: resources/testharness.js [hunk 2]
```diff
@@ -568,7 +568,21 @@ policies and contribution forms [3].
...
+                var msg = "Test named \"" + test_name +
+                    "\" inappropriately returned a value";
+
+                try {
+                    if (value && typeof value.then === 'function') {
+                        msg += ", consider using `promise_test` instead";
+                    }
+                } catch (err) {}
+
+                tests.status.status = tests.status.ERROR;
```
Comment (jgraham) [inline resources/testharness.js:582]: `tests.set_status()`

File: resources/testharness.js [hunk 3]
```diff
@@ -568,7 +568,21 @@ policies and contribution forms [3].
         var test_name = name ? name : test_environment.next_default_test_name();
         var test_obj = new Test(test_name, properties);
         if (func) {
-            test_obj.step(func, test_obj, test_obj);
+            var value = test_obj.step(func, test_obj, test_obj);
```
Comment (jgraham) [inline resources/testharness.js:571]: So we're only doing this for the initial step, not any later ones. Can we have a comment to explain why?

File: resources/testharness.js [hunk 4]
```diff
@@ -568,7 +568,21 @@ policies and contribution forms [3].
...
+            if (value !== undefined) {
+                var msg = "Test named \"" + test_name +
+                    "\" inappropriately returned a value";
```
Comment (jgraham) [inline resources/testharness.js:574]: I don't think it's clear what it means for the test to return a value.

**Review by jgraham — CHANGES_REQUESTED** [review]
Have we verified that this doesn't cause any breakage?

File: resources/testharness.js (hunk 3 above)
Comment (stephenmcgruer) [reply resources/testharness.js:572]: Added, PTAL.

File: resources/testharness.js (hunk 2 above)
Comment (stephenmcgruer) [reply resources/testharness.js]: Done.

File: resources/testharness.js (hunk 4 above)
Comment (stephenmcgruer) [reply resources/testharness.js]: I agree! I've reworked the wording here, please let me know if it reads better.

**Review by stephenmcgruer — COMMENTED** [review]
> Have we verified that this doesn't cause any breakage?

As per [this comment](https://github.com/web-platform-tests/wpt/pull/25703#issuecomment-698434806), I've run two Chrome Dev trigger runs. The first turned up ~6 cases I had previously missed, once those were fixed the second appears to be clean.

It's possible that we have some test that is returning a value only on not-Chrome, but it seems unlikely to me? Happy to do trigger runs on other browsers if you want.

**Comment by stephenmcgruer** [comment]
Given the change I had to make to `resources/test/tests/unit/exceptional-cases.html`, I wonder if we should relax this to only care about things that return then-ables. On the other hand, no real test does anything like that unittest does, so maybe its fine.

**Review by jgraham — APPROVED** [review]

**Pull request merged**